cpp/tests/binaryop/assert-binops.h

```diff
@@ -31,30 +31,31 @@ namespace cudf {
 namespace test {
 namespace binop {
 
-namespace {
-
-template <typename T, std::enable_if_t<std::is_floating_point<T>::value>* = nullptr>
-bool test_equality(T lhs, T rhs)
-{
-  constexpr int ulp = 4;  // value taken from google test
-  return !(std::abs(lhs - rhs) > std::numeric_limits<T>::epsilon() * std::abs(lhs + rhs) * ulp &&
-           std::abs(lhs - rhs) >= std::numeric_limits<T>::min());
-}
-
-template <typename T, std::enable_if_t<!std::is_floating_point<T>::value>* = nullptr>
-bool test_equality(T lhs, T rhs)
-{
-  return lhs == rhs;
-}
-
-}  // namespace
+// This comparator can be used to compare two values that are within a certain threshold.
+// This is typically used to compare floating point values computed on CPU and GPU which is
+// expected to be *near* equal, or when computing large numbers can yield ULP errors
+template <typename TypeOut>
+struct NearEqualComparator {
+  double tolerance_;
+
+  NearEqualComparator(double max_diff) : tolerance_(max_diff) {}
```

Comment:
This shouldn't be a fixed tolerance comparison. Always use ULP https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/

Reply:
done